### DISCUSSION TITLE
Router query not defined at first render? (client side)

### DISCUSSION THREAD

**Discussion by ebidrey** [vercel/next.js#14657] [Help] [answered]
Hi community,

I'm loading this url **http://localhost/search?q=pattern** and when checking the router.query, it shows me an empty object **{}**, but it should be showing the following object **{'q': 'pattern'}**. The object becomes correct in the later checks.

The matter is that I need to make an api request in client side, mapping this url parameters to API parameters as follows:
**`http://localhost/search?q=pattern` =>  /api?query=pattern**

I have done a test trying the same in server side using getInitialProps and works like a charm. But is not my objective.

Is there a way to accomplish that?

My simplified code is this:

```
import useSWR from 'swr'
import fetcher from '../lib/fetcher'
import { useRouter } from 'next/router'


function Search(props) {

    const router = useRouter()
    // first log in console is {}
    console.log(router.query) 
    //  buildParams returns "" for {}, otherwise returns param1=<value>&param2=<value>... and so on
    const paramsString = buildParams(router.query)
    const { data, error } = useSWR(`/api?${paramsString}`, fetcher,)

    if (error) return <div>failed to fetch</div>
    if (!data) return <div>loading...</div>
    return <div>fetched correctly</div>
}

export default Search
```

And what my console shows:
<img width="181" alt="Screen Shot 2020-06-28 at 13 36 49" src="https://user-images.githubusercontent.com/35265268/85954796-fb68fe00-b947-11ea-9cbb-1bc5ed2f4689.png">

EDIT:
Forgot to mention that the request to the api is being made twice, one without params, and the second correctly and the second cancels the first while is in pending status.

Thanks in advance

**Comment by jamesmosier** [comment, marked as answer]
This is expected [per the docs](https://nextjs.org/docs/routing/dynamic-routes#caveats). Next triggers an update to the router after hydration. 

You could use swr's [dependent fetching feature](https://swr.vercel.app/#dependent-fetching) to wait until the query parameters are available.


```js
const { data } = useSWR(() => {
  if (!router.query.q) throw new Error('q not available yet')
  return `/api?${paramsString}`
}))
```

**Reply by ebidrey** [reply]
First of all, thanks for your response.

My problem is that q parameter is optional (all others too), so the validation will throw this error when the parameter doesn't come for empty default searches. Maybe there are another way to check if the parameter is ready to be used?

Thanks in advance.

**Reply by jamesmosier** [reply]
Is there another reproduction you could share? I just tested this out with:

* `pages/search.js`
* http://localhost:3000/search?abc=123
* Next.js `9.4.4`

```jsx
import {useRouter} from 'next/router'

export default function SearchPage() {
  const router = useRouter();
  console.log(router.query)

  return (
    <div>
      search page
    </div>
  );
}

```

and saw the console log was correct on the first render (`{abc: "123"}`).

**Reply by ebidrey** [reply]
I've just run your code in the same version of next, and I get:

<img width="172" alt="Screen Shot 2020-06-28 at 20 30 22" src="https://user-images.githubusercontent.com/35265268/85962356-51f12f00-b97e-11ea-928c-3422f98f5211.png">

What react version are you running?

**Reply by ebidrey** [reply]
I've noticed that router.asPath has the correct queryString at mounting time, and I can accomplish the task by parsing it by myself. It's not the prettiest approach, sadly 😞

**Reply by jamesmosier** [reply]
Ah gotcha. And yeah sorry, I forgot the project I was testing in had a custom `_app.js` with `getInitialProps` in it which forces the page to be server rendered instead of statically generated.

**Reply by ebidrey** [reply]
Thanks a lot for your helpo @jamesmosier

**Reply by raine** [reply]
So what's the solution here? How does one check if `router.query.q` is ready to be used? Is there a way to check if hydration is complete?

> I've noticed that router.asPath has the correct queryString at mounting time, and I can accomplish the task by parsing it by myself. It's not the prettiest approach, sadly 😞

Yeah, I'm thinking of doing the same thing.

**Reply by Moustafa3attia** [reply]
Pretty much what I was looking for. Thanks man @jamesmosier

**Comment by ksmith0813** [comment]
Hi,

I ran into this and initially did an over complicated solution, then simplified to this:

'GET_STUFF' is a graphQL query in this case.

```
import { useRouter } from 'next/router';
import useSWR from 'swr';

const YouComponent = () => {
    const router = useRouter();
    const search = router?.query?.q;
    const { data } = useSWR(search && [GET_STUFF, { search }]);

    return (
        <div>{your stuff goes here 😄}</div>
    );
}
```

Regards,
KS
